0-module-and-module-source.emu

```diff
@@ -537,27 +569,48 @@ location: https://tc39.es/proposal-compartments/
         </ul>
 
         <emu-clause id="sec-module">
-            <h1>Module ( _moduleSource_, _referral_, _importHook_ [, _importMeta_ ] )</h1>
+            <h1>Module ( _moduleSource_, _handler_ [, _importMeta_ ] )</h1>
             <p>When the `Module` function is called with the arguments _moduleSource_, _referral_, _importHook_ and _importMeta_, the following steps are taken:</p>
             <emu-alg>
                 1. If NewTarget is *undefined*, throw a *TypeError* exception.
                 1. If Type(_moduleSource_) is not Object, throw a *TypeError* exception.
-                1. If IsCallable(_importHook_) is *false*, throw a *TypeError* exception.
-                1. If _importMeta_ is *undefined*, then
-                  1. Set _importMeta_ to *null*.
-                1. Else, if Type(_importMeta_) is not Object, then
-                  1. Throw a *TypeError* exception.
+                _importHook_
                 1. Perform ? RequireInternalSlot(_moduleSource_, [[ModuleSource]]).
-                1. Let _O_ be ? OrdinaryCreateFromConstructor(NewTarget, *"%Module.prototype%"*, « [[Module]], [[ModuleSourceInstance]], [[ImportHook]] »).
                 1. Let _moduleRecord_ to ! CreateModuleRecord(_moduleSource_.[[ModuleSource]]).
-                1. Set _moduleRecord_.[[ModuleInstance]] to _O_.
-                1. Set _moduleRecord_.[[ImportMeta]] to _importMeta_.
-                1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _importHook_ and _referral_ and performs the following steps when called
-                  1. Assert: Type(_specifier_) is String.
-                  1. return ? Call(_importHook_, *undefined*, « _specifier_, _referral_ »).
-                1. Set _O_.[[Module]] to _moduleRecord_.
+                1. If Type(_handler_) is Object, then
+                  1. If _handler_ is *null, throw a TypeError exception.
+                  1. Let _importHook_ be ? GetMethod(handler, "importHook").
+                  1. If _importHook_ is *undefined*, then
+                    1. Let _importHookClosure_ be a new Abstract Closure with parameters (_specifier_) that captures _moduleRecord_ and performs the following steps when called
+                      1. Assert: Type(_specifier_) is String.
+                      1. NOTE: Default import hook.
+                      1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
+                      1. Perform HostImportModuleDynamically(_moduleRecord_, _specifier_, _promiseCapability_).
```

Comment:
@kriskowal @nicolo-ribaudo this first argument might be problematic because it is a record from an instance, the host might or might not know about it. Ultimately, the host can detect (from looking at the hostDefined) whether or not the referral is good, or they should assume global scope for the resolution and metas. Let's chat about it.